JavaScriptSPA/authRedirect.js

```diff
@@ -0,0 +1,69 @@
+// Create the main myMSALObj instance
+// configuration parameters are located at authConfig.js
+const myMSALObj = new Msal.UserAgentApplication(msalConfig);
+
+let accessToken;
+// Register Callbacks for Redirect flow
+myMSALObj.handleRedirectCallback(authRedirectCallBack);
+
+function authRedirectCallBack(error, response) {
+    if (error) {
+        console.log(error);
+    } else {
+        if (response.tokenType === "id_token") {
+            console.log('id_token acquired at: ' + new Date().toString());
+            myMSALObj.getAccount();
+            getTokenRedirect(tokenRequest);
+        } else if (response.tokenType === "access_token") {
+            console.log('access_token acquired at: ' + new Date().toString());
+            accessToken = response.accessToken;
+            logMessage("Request made to Web API:")
+            callApiWithAccessToken(apiConfig.webApi, accessToken)
+        } else {
+            console.log("token type is:" + response.tokenType);
+        }
+    }
+}
+
+// Redirect: once login is successful and redirects with tokens, update UI
+if (myMSALObj.getAccount()) {
+    updateUI();
+}
+
+function signIn() {
+    myMSALObj.loginRedirect(loginRequest)
+}  
+
+
+// sign-out the user
+function logout() {
+    // Removes all sessions, need to call AAD endpoint to do full logout
+    myMSALObj.logout();
+}
+
+// This function can be removed if you do not need to support IE
+function getTokenRedirect(request) {
+return myMSALObj.acquireTokenSilent(request)
+    .then((response) => {
+        if (response.accessToken) {
+            accessToken = response.accessToken
+            logMessage("Request made to Web API:")
+            callApiWithAccessToken(apiConfig.webApi, accessToken)
+        }
+    }).catch(error => {
+        console.log("silent token acquisition fails. acquiring token using redirect");
+        // fallback to interaction when silent call fails
+        return myMSALObj.acquireTokenRedirect(request)
+    });
+}
+
+
+// calls the resource API with the token
```

Comment:
So here is the thing ... @mmacy had the feedback that API access code needs to be separated from token acquisition! By having `callApi()` in the same file as `getToken` we are defeating that purpose. Is there a way we can pass tokens from a file to the other and restrict `callApi()` to API access only? We can do that in index.html also.

Same applies for VanillaJS samples (non B2C I mean).

cc @mmacy to add his feedback.

Reply:
@sameerag @derisen I haven't yet had a chance to run through the tutorials with the new sample, but essentially, the user needs to be able to:

1. Tutorial 1: Sign in to the application
1. Tutorial 2: Access the API

At the completion of tutorial 1, the user won't yet have setup or configured the API pieces, so sign-in needs to work on its own. Then, at the end of tutorial 2, after they've set up the API pieces, they hit the "access API" button and that's when it calls the API.

Reply:
Popup flow works just fine without anything api related. I've moved `callAPI()` (it's badly named, all it does is to get an access token and pass it to the `callApiWithAccessToken()` function) back to `api.js`, then simply didn't include those api related files in the loadup, and it authenticates all fine (`authPopup.js` just gets an id token). So there should be no problem for tutorials that way.

The issue is with the redirect flow. I call `callApiWithAccessToken()` in the redirect callback function, so when the page loads with the access token it calls the API.
 
To fully seperate these logics I can't think of a way right now other than having a global for accessToken in auth.js files. Then we would just pass it in the html like: 

`<button onclick="callApi(accessToken)">Call API</button>`

I guess it would be best to figure this out after we sync @sameerag ?

Reply:
From the tutorial flow standpoint, as long as it doesn't throw an error after sign-in but before they've finished tutorial 2 and set up the API bits, we're golden.

Reply:
@derisen Apologies on the delay on sync up. Does today work for you? Please message me. @mmacy we will have an answer to you by tomorrow on the status.

Reply:
@sameerag no worries! sure thing, let me know time

Reply:
@sameerag @derisen No worries, and no hurry - I have *plenty* on my plate. :wink:

Reply:
@mmacy an update: We talked (and a few more msal js team members) and except finalizing the `redirect` code,  we have reached to a conclusion on most questions. We will try to close the pending AI next week. Thanks.

Reply:
@mmacy @sameerag I think we are good to go now. From the tutorials perspective, here's what I think:

The default flow for the tutorial should be Popup. We also provide an optional `authRedirect.js` file for people who want try it, but it should be tried after the tutorials. 

For part 2: "Grant access to API", the user will do these:
- create `api.js` file to call the api
- crate `apiConfig.js` file to enter the api parameters
- go back to `authPopup.js` and add the function `passTokenToApi()`
- go back to `authConfig.js` file to add `tokenRequest` scopes
- go back to `index.html` file to to add the "Call API" button

Even using `authRedirect.js` will not break anything, they will still login all fine, its just that the code has more references to `callApiWithAccessToken` function due to the nature of redirect flow (so its more tedious to go back and add this line where its needed), but since there'll be no "Call API" button on the user interface, it will not run (and even if it runs, nothing will break, just a white error).

Reply:
Sounds good me. Excellent improvement overall! cc @ktsakas @jmprieur